daemon/list_test.go

```diff
@@ -79,6 +85,49 @@ func containerListContainsName(containers []*containertypes.Summary, name string
 	return false
 }
 
+func TestList(t *testing.T) {
+	db, err := container.NewViewDB()
+	assert.NilError(t, err)
+	d := &Daemon{
+		containersReplica: db,
+	}
+
+	// test list with no containers
+	containerList, err := d.Containers(context.Background(), &containertypes.ListOptions{})
+	assert.NilError(t, err)
+	assert.Assert(t, is.Len(containerList, 0))
+
+	// test list of one container
+	one := setupContainerWithName(t, "a1", d)
+	containerList, err = d.Containers(context.Background(), &containertypes.ListOptions{})
+	assert.NilError(t, err)
+	assert.Assert(t, is.Len(containerList, 1))
+	assert.Assert(t, containerListContainsName(containerList, one.Name))
+
+	// test list with random number of containers
+	db2, err := container.NewViewDB() // new DB to ignore prior containers
+	assert.NilError(t, err)
+	d = &Daemon{
+		containersReplica: db2,
+	}
+
+	// start a random number of containers (between 0->256)
+	num := rand.Intn(256)
+	containers := make([]*container.Container, num)
+	for i := range num {
+		name := fmt.Sprintf("cont-%d", i)
+		containers[i] = setupContainerWithName(t, name, d)
+	}
+
+	// list them and verify correctness
+	containerList, err = d.Containers(context.Background(), &containertypes.ListOptions{})
+	assert.NilError(t, err)
+	assert.Assert(t, is.Len(containerList, num))
+	for i := range num {
+		assert.Assert(t, containerListContainsName(containerList, containers[i].Name))
+	}
```

Comment:
Curious; I'm guessing the intent here is to catch possible race conditions where containers may be dropped? Do you have examples where this caught issues?

Curious because I don't thing we have tests taking this approach 😅

Reply:
Yes exactly: I added the randomness in the number of containers to test that we are not dropping any containers due to a race condition in the worker goroutines.